src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,267 @@
+using ..Turing
+using ..Bijectors
+using LinearAlgebra
+
+import ..AbstractMCMC: AbstractSampler
+import ..DynamicPPL
+import ..DynamicPPL: Model, AbstractContext, VarInfo, AbstractContext, VarName,
+    _getindex, getsym, getfield, settrans!,  setorder!,
+    get_and_set_val!, istrans, tilde, dot_tilde
+import Optim
+import Optim: optimize
+import NamedArrays
+import ..ForwardDiff
+import StatsBase
+import Printf
+
+struct MLE end
+struct MAP end
+
+"""
+    OptimizationContext{C<:AbstractContext} <: AbstractContext
+
+The `OptimizationContext` transforms variables to their constrained space, but
+does not use the density with respect to the transformation. This context is
+intended to allow an optimizer to sample in R^n freely.
+"""
+struct OptimizationContext{C<:AbstractContext} <: AbstractContext
+	context::C
+end
+
+# assume
+function DynamicPPL.tilde(ctx::OptimizationContext{<:LikelihoodContext}, spl, dist, vn::VarName, inds, vi)
+    r = vi[vn]
+	return r, 0
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, spl, dist, vn::VarName, inds, vi)
+    r = vi[vn]
+    return r, Distributions.logpdf(dist, r)
+end
+
+# observe
+function DynamicPPL.tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vi)
+    return 0
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, sampler, dist, value, vi)
+    return Distributions.logpdf(dist, value)
+end
+
+# dot assume
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:LikelihoodContext}, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    return vi[vn], 0
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = vi[vn]
+    return r, loglikelihood(dist, r)
```

Comment:
I think we should get rid of the `sum(logpdf.(dist, r))` calls but as I said I assume that just changing them to `loglikelihood` will break AD and be problematic if `dist` is an array of distributions. Maybe we should do that in a separate PR, I guess by introducing a separate `sum_logpdf` in DistributionsAD (to avoid having to commit type piracy when defining `loglikelihood(dist::AbstractArray{<:Distribution}, x)`) with optimized adjoints for AD and a fallback to `loglikelihood(dist::Distribution, x)`.

Reply:
Agreed. I'm sure we'll see how this'll fit with AD as people test it out on more complex models.